api/envoy/config/core/v3/base.proto

```diff
@@ -226,6 +226,8 @@ message Metadata {
   // Key is the reverse DNS filter name, e.g. com.acme.widget. The envoy.*
   // namespace is reserved for Envoy's built-in filters.
   map<string, google.protobuf.Struct> filter_metadata = 1;
+
```

Comment:
The question here is shall we deprecate `filter_metadata`? When we changed filter configs from Struct to Any we deprecated Struct because Any can have Struct in it.

Reply:
let's mark it as deprecated. 

I think Oneof will make it harder to migrate.
For now we need both to have a smooth migration: let the metadata impl first look at typed_filter_metadata and fallback to typed_metadata when it's not populated. This way we can decouple dataplane with control plane.

Reply:
Yeah don't make it oneof, it breaks the compatibility on generated codes and violates our API policy.

Reply:
Using Oneof is infeasible here. For Oneof fields, "You can add fields of any type, except map fields and repeated fields." See http://go/proto3#oneof

Reply:
+1 on deprecate untyped (and ideally make typed capable of recognizing `Struct-inside-of-Any`).

Reply:
The main **present** issue with transitioning to using `Struct-inside-of-Any` is that the [Metadata utility class](https://github.com/envoyproxy/envoy/blob/7cd615c262dc1b871812209e37c15a49daea5173/source/common/config/metadata.h#L37) has static methods only. It will be inefficient to use with the `Struct-inside-of-Any` as there is no way to cache parsed Struct. The Metadata utility class is used heavily within Envoy and also by proprietary filters, so changing its API will be challenging.
One potential way forward is to make current implementation of the Metadata class inefficient with respect to `Struct-inside-of-Any` and mark it as deprecated to prevent further use. And then add a new class that can cache parsed Struct from Any.

Reply:
Yeah, I think a `TypedMetadata` class would make sense.

Reply:
Thanks for the comments. Based on the discussion,  I put Struct as deprecated, and added Any. Please take a look the latest code change and let me know whether this look good to you.

Reply:
With Struct field being deprecated, this pre_submit check scripts: " ci/run_envoy_docker.sh 'ci/do_ci.sh bazel.compile_time_options' ", is complaining below and failed the check constantly:

2021-05-07T23:34:58.2181735Z C++ exception with description "type envoy.config.core.v3.Metadata Using deprecated option 'envoy.config.core.v3.Metadata.filter_metadata' from file base.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/version_history/version_history for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override." thrown in the test body.
2021-05-07T23:34:58.2184225Z [  FAILED  ] OmitHostsRetryPredicateTest.PredicateTest (22 ms)

This happend on a lot of test cases, and one scripts run only report some of them. Once some test cases are failed as above, I just use DEPRECATED_FEATURE_TEST() to guard them. Another scripts run will report new ones.   I have below questions:

1) Should I continue this catch and patch style to mark all those test cases which exercising Metadata:Struct with  DEPRECATED_FEATURE_TEST()?  
2) Is there a way to let the scripts run to the end and report all the test cases which exercising Metadata:Struct ? That way I can patch once for all.
3) Is there a timing pressure to commit Any in the metadata proto? If there is, should we first commit the Any without flag Struct as deprecated?  That way we deprecate Struct  and fix all the text cases in a separate PR.

Please let me know your opinion.

Thanks!